### PR TITLE
T10595

### PR DESCRIPTION
**Pull request by ptomato**
https://phabricator.endlessm.com/T10595

### CODE REVIEW

File: test/utils.js [hunk 1]
```diff
@@ -20,9 +20,37 @@
...
+const EosShard = imports.gi.EosShard;
+
 function getTestFile(fn) {
     let datadir = GLib.getenv('G_TEST_SRCDIR');
     if (!datadir)
         datadir = '';
     return Gio.File.new_for_path(GLib.build_filenamev([datadir, 'test/data', fn]));
 }
+
+function writeTestShard(shard_path) {
```
Comment (magcius) [inline test/utils.js]: I actually prefer to repeat myself in tests rather than deduplicate, because it makes tests easier to read and debug IMO.

**Comment by magcius** [comment]
It would be a lot better if we didn't depend on this behavior. From the GAsyncInitiable docs:

```
Implementations may also support cancellation. If cancellable is not NULL, then initialization can be cancelled by triggering the cancellable object from another thread. If the operation was cancelled, the error G_IO_ERROR_CANCELLED will be returned.

As with GInitable, if the object is not initialized, or initialization returns with an error, then all operations on the object except g_object_ref() and g_object_unref() are considered to be invalid, and have undefined behaviour.```

So the "correct" way of implementing this would be to recreate the object from outside. This is admittedly confused by the example code immediately under those paragraphs which does correctly support `init_async` on a failed object.

**Comment by magcius** [comment]
This is also not allowed according to the docs:

>  Implementations of this method must be idempotent, i.e. multiple calls to this function with the same argument should return the same results. Only the first call initializes the object, further calls return the result of the first call. This is so that it's safe to implement the singleton pattern in the GObject constructor function.

**Comment by ptomato** [comment]
This seems like overly restrictive behaviour for GAsyncInitable with respect to GCancellable, but the docs bear it out.

I'll make the fix in ekn-search-provider instead. Yesterday I was thinking that might still have a race condition somewhere, but now I can't remember where I saw it, and it seems sound to me.

**Comment by magcius** [comment]
Yeah, the docs are also in direct conflict with each other -- "init_async needs to be idempotent, even after failures" and "anything other than ref/unref is undefined behavior after failures".

In any case, we never really want to cancel a shard load, since we'll use it eventually.

**Pull request closed without merging**